### PR TITLE
Draft: max detour constraint drt

### PR DESCRIPTION
**Pull request by nkuehnel**

### CODE REVIEW

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+	private boolean violatesDetour(InsertionGenerator.Insertion insertion, VehicleEntry vEntry, DrtRequest drtRequest) {
+		// Check if the max travel time constraint for the newly inserted request is violated
+		double rideDuration = insertion.dropoff.newWaypoint.getArrivalTime() - insertion.pickup.newWaypoint.getDepartureTime();
+		if(drtRequest.getMaxRideDuration() < rideDuration) {
+			return true;
+		}
+
+		final int pickupIdx = insertion.pickup.index;
+		Map<Id<Request>, Double> pickUps = new HashMap<>(ongoingRequests);
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java:43]: What is the point of copying this map?

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.matsim.contrib.dvrp.passenger.PassengerDroppedOffEventHandler;
+import org.matsim.contrib.dvrp.passenger.PassengerPickedUpEvent;
+import org.matsim.contrib.dvrp.passenger.PassengerPickedUpEventHandler;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class MaxDetourInsertionCostCalculator implements InsertionCostCalculator, PassengerPickedUpEventHandler, PassengerDroppedOffEventHandler {
+
+	private final Map<Id<Request>, Double> ongoingRequests = new HashMap<>();
```
Comment (michalmac) [inline contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java:19]: This map can be read and written at the same time (events handling is executed in parallel to optimisation)

File: contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java (hunk 1 above)
Comment (michalmac) [reply contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/MaxDetourInsertionCostCalculator.java:43]: I think I understand the idea. However, I think copying the map and then adding even more stuff to it is a bit too heavy.

**Comment by luchengqi7** [comment]
Hi Nico, it seems that our hack does not work with the current insertion logic. The arrival time of the WayPoint is not implemented. 
https://github.com/matsim-org/matsim-libs/blob/5695f6309935b71d5ccb59fcbf8b13a23dbc0811/contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/Waypoint.java#L260

I have discussed with Michal (@michalmac), and now we think it may be better to have some more fundamental implementation, rather than changing the current basic structure of the insertion. 

We can have a discussion after you return from your vacation.

**Comment by luchengqi7** [comment]
Hi Nico,

How are you? Hope everything is running smoothly :) 

May I know if there is any update from your side on this topic? Shall we go ahead to change the structure of the insertion logic, or perhaps we can think of an alternative quick fix? 

Best regards,
Chengqi Lu

**Comment by sebhoerl** [comment]
Just as comment, it might also be worth it thinking about adding this as a constraint here :)
https://github.com/matsim-org/matsim-libs/pull/2947

**Comment by michalmac** [comment]
> > Just as comment, it might also be worth it thinking about adding this as a constraint here :) #2947
> 
> That's a great point! Thanks a lot, Sebastian!

My suggestion would be to use the max detour time to pre-compute slack times before generating insertions. Otherwise, if we only post-process insertions, we may experience an "unexplainable" increase in computation time when changing from max travel time to max detour time.

**Comment by nkuehnel** [comment]
Hey @luchengqi7 ! sorry for my late reply, it was a bit busy.

Maybe we can have a chat next week? I'll write a mail

**Comment by michalmac** [comment]
I just pushed a commit that reverts the changes in `AbstractRoute` (this was not intended) and instead introduces them in `DrtRoute`.

@nkuehnel @luchengqi7 I wanted to bring this up for your next week sync: `DrtConfigGroup.maxAbsoluteDetour` is used for the total travel time (wait+ride). I guess we need something else for the max detour wrt ride time? Right now this logic is not correct (see the first argument in both `min()` methods):
```java
	static double getMaxTravelTime(DrtConfigGroup drtCfg, double unsharedRideTime) {
		return Math.min(unsharedRideTime + drtCfg.maxAbsoluteDetour, drtCfg.maxTravelTimeAlpha * unsharedRideTime + drtCfg.maxTravelTimeBeta);
	}

	static double getMaxRideTime(DrtConfigGroup drtCfg, double unsharedRideTime) {
		return Math.min(unsharedRideTime + drtCfg.maxAbsoluteDetour, drtCfg.maxDetourAlpha * unsharedRideTime + drtCfg.maxDetourBeta);
	}
```

**Comment by luchengqi7** [comment]
Hi @nkuehnel, today I have discussed with Michal, and he is happy about this new idea. Since there are some other changes and conflicts in this branch, shall I open a new PR in the current matsim-lib (upstream)?

**Review by luchengqi7 — APPROVED** [review]

**Comment by luchengqi7** [comment]
The proposed changes in this PR have been merged to the master branch by PR #2997.

**Pull request closed without merging**